src/hdchain.cpp

```diff
@@ -0,0 +1,211 @@
+{
+    SecureString ssMnemonicTmp = ssMnemonic;
+
+    if (fUpdateID) {
+        // can't (re)set mnemonic if seed was already set
+        if (!IsNull())
+            return false;
+
+        // empty mnemonic i.e. "generate a new one"
+        if (ssMnemonic.empty()) {
+            ssMnemonicTmp = CMnemonic::Generate(256);
+        }
+        // NOTE: default mnemonic passphrase is an empty string
+
+        // printf("mnemonic: %s\n", ssMnemonicTmp.c_str());
+        if (!CMnemonic::Check(ssMnemonicTmp)) {
+            throw std::runtime_error(std::string(__func__) + ": invalid mnemonic: `" + std::string(ssMnemonicTmp.c_str()) + "`");
+        }
+
+        CMnemonic::ToSeed(ssMnemonicTmp, ssMnemonicPassphrase, vchSeed);
+        id = GetSeedHash();
+    }
+
+    vchMnemonic = SecureVector(ssMnemonicTmp.begin(), ssMnemonicTmp.end());
+    vchMnemonicPassphrase = SecureVector(ssMnemonicPassphrase.begin(), ssMnemonicPassphrase.end());
+
+    return !IsNull();
+}
+
+bool CHDChain::GetMnemonic(SecureVector& vchMnemonicRet, SecureVector& vchMnemonicPassphraseRet) const
+{
+    // mnemonic was not set, fail
+    if (vchMnemonic.empty())
+        return false;
+
+    vchMnemonicRet = vchMnemonic;
+    vchMnemonicPassphraseRet = vchMnemonicPassphrase;
+    return true;
+}
+
+bool CHDChain::GetMnemonic(SecureString& ssMnemonicRet, SecureString& ssMnemonicPassphraseRet) const
+{
+    // mnemonic was not set, fail
+    if (vchMnemonic.empty())
+        return false;
+
+    ssMnemonicRet = SecureString(vchMnemonic.begin(), vchMnemonic.end());
+    ssMnemonicPassphraseRet = SecureString(vchMnemonicPassphrase.begin(), vchMnemonicPassphrase.end());
+
+    return true;
+}
+
+bool CHDChain::SetSeed(const SecureVector& vchSeedIn, bool fUpdateID)
+{
+    vchSeed = vchSeedIn;
+
+    if (fUpdateID) {
+        id = GetSeedHash();
+    }
+
+    return !IsNull();
+}
+
+SecureVector CHDChain::GetSeed() const
+{
+    return vchSeed;
+}
+
+uint256 CHDChain::GetSeedHash()
+{
+    return Hash(vchSeed.begin(), vchSeed.end());
+}
+
+void CHDChain::DeriveChildExtKey(uint32_t nAccountIndex, bool fInternal, uint32_t nChildIndex, CExtKey& extKeyRet)
+{
+    // Use BIP44 keypath scheme i.e. m / purpose' / coin_type' / account' / change / address_index
+    CExtKey masterKey;              //hd master key
+    CExtKey purposeKey;             //key at m/purpose'
+    CExtKey cointypeKey;            //key at m/purpose'/coin_type'
+    CExtKey accountKey;             //key at m/purpose'/coin_type'/account'
+    CExtKey changeKey;              //key at m/purpose'/coin_type'/account'/change
+    CExtKey childKey;               //key at m/purpose'/coin_type'/account'/change/address_index
+
+    masterKey.SetMaster(&vchSeed[0], vchSeed.size());
+
+    // Use hardened derivation for purpose, coin_type and account
+    // (keys >= 0x80000000 are hardened after bip32)
+
+    // derive m/purpose'
+    masterKey.Derive(purposeKey, 44 | 0x80000000);
```

Comment:
nit: what is the meaning of 44 here? Better to avoid magic numbers.

Reply:
its the purpose level,as commented by random zebra on hdchain.cpp conversation.